server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterInfoSimulator.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.DiskUsage;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class ClusterInfoSimulator {
+
+    private final Map<String, DiskUsage> leastAvailableSpaceUsage;
+    private final Map<String, DiskUsage> mostAvailableSpaceUsage;
+    private final Map<String, Long> shardSizes;
+    private final Map<ShardId, Long> shardDataSetSizes;
+    private final Map<ShardRouting, String> routingToDataPath;
+    private final Map<ClusterInfo.NodeAndPath, ClusterInfo.ReservedSpace> reservedSpace;
+
+    public ClusterInfoSimulator(ClusterInfo clusterInfo) {
+        this.leastAvailableSpaceUsage = new HashMap<>(clusterInfo.getNodeLeastAvailableDiskUsages());
+        this.mostAvailableSpaceUsage = new HashMap<>(clusterInfo.getNodeMostAvailableDiskUsages());
+        this.shardSizes = new HashMap<>(clusterInfo.shardSizes);
+        this.shardDataSetSizes = new HashMap<>(clusterInfo.shardDataSetSizes);
+        this.routingToDataPath = new HashMap<>(clusterInfo.routingToDataPath);
+        this.reservedSpace = new HashMap<>(clusterInfo.reservedSpace);
+    }
+
+    public void simulate(ShardRouting shard) {
+        assert shard.initializing();
+
+        var size = getEstimatedShardSize(shard);
+        if (size != null && size > 0) {
+            if (shard.relocatingNodeId() != null) {
+                // relocation
+                increaseTargetDiskUsage(shard.relocatingNodeId(), size);
+                decreaseSourceDiskUsage(shard.currentNodeId(), size);
+                // TODO update shard data path?
```

Comment:
`routingToDataPath` is keyed by a ShardRouting. If we want to update it as well then we need to also pass the state of the shard after it was initialized here.